### PR TITLE
Kie MacPherson

### PR DESCRIPTION
**Pull request by Reverb-sudo**

### CODE REVIEW

File: brevets/acp_times.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+       brevet_dist_km: number, nominal distance of the brevet
+           in kilometers, which must be one of 200, 300, 400, 600,
+           or 1000 (the only official ACP brevet distances)
+       brevet_start_time:  An arrow object
+    Returns:
+       An arrow object indicating the control open time.
+       This will be in the same time zone as the brevet start time.
+    """
+    minutes = 0
+    TRIGGERS = [(200,200,34),(400,200,32),(600,200,30),(1000,400,28),(1300,300,26)]
```
Comment (freddy-lopez01) [inline brevets/acp_times.py:26]: Love the name of this list haha!

File: brevets/acp_times.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+following rules described at https://rusa.org/octime_acp.html
+and https://rusa.org/pages/rulesForRiders
+"""
+import arrow
+#  You MUST provide the following two functions
+#  with these signatures. You must keep
+#  these signatures even if you don't use all the
+#  same arguments.
+#
+def open_time(control_dist_km, brevet_dist_km, brevet_start_time):
```
Comment (freddy-lopez01) [inline brevets/acp_times.py:13]: This function is very concise and efficient compared to others and even my own! nice work

File: brevets/acp_times.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        if control_dist_km > km:
+            minutes += (l / max) * 60
+        else:
+            minutes += ((control_dist_km-(km-l))/max) * 60
+            break
+    rounded = round(minutes)
+    return brevet_start_time.shift(minutes=rounded)
+
+
+def close_time(control_dist_km, brevet_dist_km, brevet_start_time):
```
Comment (freddy-lopez01) [inline brevets/acp_times.py:40]: Something I noticed was, does your close_time() take into account of the possibility for a control brevet to be 20% longer than the total brevet itself. Its one of those edge cases that we are supposed to support in out function. I dont see it here so you might want to look into that but if it is and im just not seein it my bad

File: brevets/flask_brevets.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+###############
+@app.route("/_calc_times")
+def _calc_times():
+    """
+    Calculates open/close times from miles, using rules
+    described at https://rusa.org/octime_alg.html.
+    Expects one URL-encoded argument, the number of miles.
+    """
+    app.logger.debug("Got a JSON request")
+    km = request.args.get('km', 999, type=float)
```
Comment (freddy-lopez01) [inline brevets/flask_brevets.py:53]: good variable naming conventions

File: brevets/flask_brevets.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    # and control distance is fixed to 200
+    # You should get these from the webpage!
+    open_time = acp_times.open_time(km, brevet_length, arrow.get(start_time)).format('YYYY-MM-DDTHH:mm')
+    close_time = acp_times.close_time(km, brevet_length, arrow.get(start_time)).format('YYYY-MM-DDTHH:mm')
+    result = {"open": open_time, "close": close_time}
+    return flask.jsonify(result=result)
+
+@app.route("/_insert", methods=["POST"])
+def _insert():
+    # start = request.args.get('start', type=str),
```
Comment (freddy-lopez01) [inline brevets/flask_brevets.py:70]: I did the same thing initially haha

File: brevets/flask_brevets.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    # app.logger.debug("kmlist={}".format(km))
+    app.logger.debug("hit insert")
+    app.logger.debug(request.args.get('start', type=str))
+
+    input_json = flask.request.json
+
+    return flask.jsonify(result=
+    brevet_insert(
+    input_json["start"],
+    #request.args.get('brevet_dist_km', type=str),
```
Comment (freddy-lopez01) [inline brevets/flask_brevets.py:84]: Not a big deal but maybe remove the commented out code that didn't end up working just to improve readability a bit

File: brevets/mypymongo.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
+import os
+from pymongo import MongoClient
+
+client = MongoClient('mongodb://' + os.environ['MONGODB_HOSTNAME'], 27017)
+db = client.mydb
+collection = db.lists
+
+
+def brevet_insert(start, brevet_dist_km, controllist):
+    print("hit brevet_insert")
```
Comment (freddy-lopez01) [inline brevets/mypymongo.py:10]: nice debugging print statements to track where you are

File: brevets/mypymongo.py [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    'controllist': controllist}
+    )
+    theid = thecollection.inserted_id
+    return str(theid)
+
+def brevet_fetch():
+    lists = collection.find().sort("_id",-1).limit(1)
+    for list in lists:
+        del list["_id"]
+        return list
```
Comment (freddy-lopez01) [inline brevets/mypymongo.py:23]: nice stylistic choice on returning the list. Makes for cleaner code

File: brevets/templates/calc.html [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+               });
+            }
+
+            $(document).ready(function(){
+             // Do the following when the page is finished loading
+
+               $("#insert").click(function(event) { //submit
+                 event.preventDefault();
+                 console.log("entered submit");
+
```
Comment (freddy-lopez01) [inline brevets/templates/calc.html:155]: Interesting and cleaner way of obtaining values from the webpage without the use of a $(".control").each(function{}) I wasnt aware of querySelectorAll but in the future ill use that

File: brevets/templates/calc.html [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+                 }
+                 console.log(lists);
+                //$.getJSON(INSERT_URL, lists);
+                $.ajax({
+                  // Expect JSON responses from our backend
+                  dataType: 'json', // expected output format (getJSON sets this as well)
+                  // Specify we're sending JSON as well
+                  contentType: "application/json",
+                  // Request method: POST
+                  type: 'POST',
```
Comment (freddy-lopez01) [inline brevets/templates/calc.html:179]: good specifying POST requests

File: brevets/tests/test_acp_times.py [hunk 11]
```diff
@@ -0,0 +1,104 @@
+"""
+Nose tests for acp_times.py
```
Comment (freddy-lopez01) [inline brevets/tests/test_acp_times.py:2]: good test cases and good variation in times

File: brevets/tests/test_acp_times.py [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+
+def test_brevet5():
+    start_time = arrow.get("2023-02-17 00:00", "YYYY-MM-DD HH:mm")
+    dist = 200
+    checkpoints = {
+        200: (start_time, start_time.shift(hours=13,minutes=30)),
+    }
+    for km, time_tuple in checkpoints.items():
+        checkpoint_open, checkpoint_close = time_tuple
+        #assert open_time(km, dist, start_time) == checkpoint_open
```
Comment (freddy-lopez01) [inline brevets/tests/test_acp_times.py:94]: May want to test edge cases regarding varying end time exceptions depending on total brevet distance

File: brevets/acp_times.py (hunk 1 above)
Comment (freddy-lopez01) [inline brevets/acp_times.py:26]: Good use of list to hold brevet distances and speeds

File: brevets/acp_times.py [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+    Returns:
+       An arrow object indicating the control open time.
+       This will be in the same time zone as the brevet start time.
+    """
+    minutes = 0
+    TRIGGERS = [(200,200,34),(400,200,32),(600,200,30),(1000,400,28),(1300,300,26)]
+    if brevet_dist_km < control_dist_km:
+        control_dist_km = brevet_dist_km
+    for trigger in TRIGGERS:
+        km, l, max = trigger
```
Comment (freddy-lopez01) [inline brevets/acp_times.py:30]: good use minimizing excessive for loops for open_time

File: brevets/acp_times.py [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+        if control_dist_km > km:
+            minutes += (l / max) * 60
+        else:
+            if control_dist_km <= 60: #oddity
+                minutes += ((control_dist_km/20) + 1) *60
+                break
```
Comment (freddy-lopez01) [inline brevets/acp_times.py:64]: Id check the edge case that involves a brevet that is 20% longer than the original brevet which gets closed at the same time as the final checkpoint

**Review by freddy-lopez01 — COMMENTED** [review]
Overall very good and readable code. As mentioned within the comments, I would check close times specially for the edge cases that involve a control brevet thats longer than the brevet itself. The test cases regardless are thought out and reasonable. Your approach to the html and the test cases is very similar to how I did but different enough to were it was helpful seeing different approaches.
